domain/domain.go

```diff
@@ -932,6 +940,91 @@ func (do *Domain) initLogBackup(ctx context.Context, pdClient pd.Client) error {
 	return nil
 }
 
+// when tidb_replica_read = 'closest-adaptive', check tidb and tikv's zone label matches.
+// if not match, disable replica_read to avoid uneven read traffic distribution.
+func (do *Domain) closestReplicaReadCheckLoop(ctx context.Context, pdClient pd.Client) {
+	defer util.Recover(metrics.LabelDomain, "closestReplicaReadCheckLoop", nil, false)
+
+	// trigger check once instantly.
+	if err := do.checkReplicaRead(ctx, pdClient); err != nil {
+		logutil.BgLogger().Warn("refresh replicaRead flag failed", zap.Error(err))
+	}
+
+	ticker := time.NewTicker(time.Minute)
+	defer func() {
+		ticker.Stop()
+		do.wg.Done()
+		logutil.BgLogger().Info("closestReplicaReadCheckLoop exited.")
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			if err := do.checkReplicaRead(ctx, pdClient); err != nil {
+				logutil.BgLogger().Warn("refresh replicaRead flag failed", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (do *Domain) checkReplicaRead(ctx context.Context, pdClient pd.Client) error {
+	// fast path
+	do.sysVarCache.RLock()
+	replicaRead := do.sysVarCache.global[variable.TiDBReplicaRead]
+	do.sysVarCache.RUnlock()
+
+	if !strings.EqualFold(replicaRead, "closest-adaptive") {
+		logutil.BgLogger().Info("closest replica read is not enabled, skip check!", zap.String("mode", replicaRead))
+		return nil
+	}
+	stores, err := pdClient.GetAllStores(ctx, pd.WithExcludeTombstone())
+	if err != nil {
+		return err
+	}
+
+	zones := make(map[string]int)
+	for _, s := range stores {
+		// skip tumbstone stores or tiflash
+		if s.NodeState == metapb.NodeState_Removing || s.NodeState == metapb.NodeState_Removed || engine.IsTiFlash(s) {
+			continue
+		}
+		for _, label := range s.Labels {
+			if label.Key == placement.DCLabelKey && label.Value != "" {
```

Comment:
will tikv be label `zone=xxx` in cloud?

Reply:
Yes.This special labels is assign either the user provide zone label or from the k8s label `topology.kubernetes.io/zone`. I'm not sure if it's better to allow the user to config the DCLabelKey instead

Reply:
but I check a cloud, the label key is `failure-domain.beta.kubernetes.io/zone`,  it is not equal to the `placement.DCLabelKey`.

Reply:
`failure-domain.beta.kubernetes.io/zone` is the deprecated k8s location label. In newer version, the label is `topology.kubernetes.io/zone`.

Reply:
I have update the logic in tidb-operator to use one of ["zone", "topology.kubernetes.io/zone", "failure-domain.beta.kubernetes.io/zone"], so it can be compatible in most case.

Reply:
Can we use `zone` directly? if the label key is too long, there is a performance issue https://github.com/tikv/pd/issues/5360. and also, it's more friendly for the user on the cloud to use placement rules.

Reply:
Yes. operator will always set the `zone` label with one of `["zone", "topology.kubernetes.io/zone", "failure-domain.beta.kubernetes.io/zone"]`'s value, so we don't need to change the logic in tidb.
See:   https://github.com/pingcap/tidb-operator/blob/7551e27bede5938a868f2a6e718ab503e60203e8/pkg/manager/member/tidb_member_manager.go#L1120